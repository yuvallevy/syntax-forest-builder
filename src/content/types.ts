export type Id = string;
export type Sentence = string;
export type NodeLabel = string;

type SliceStart = number;
type SliceEndExclusive = number;

export type StringSlice = [SliceStart, SliceEndExclusive];

export type IdMap<T> = Record<Id, T>;

export type NodeIndicatorInPlot = {
  treeId: Id;
  nodeId: Id;
};

export type TreeCommon = {
  sentence: Sentence;
}

export type NodeCommon = {
  label: NodeLabel;
};
