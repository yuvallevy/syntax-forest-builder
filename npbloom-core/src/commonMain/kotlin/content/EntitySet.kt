@file:OptIn(ExperimentalJsExport::class)

package content

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.js.ExperimentalJsExport
import kotlin.js.JsExport
import kotlin.js.JsName

/**
 * Wrapper around a set of objects that have an `id` field (all types of trees and nodes).
 * Previously a simple `Map<Id, T>` was used for fast access by ID, and objects themselves did not store their IDs.
 * EntitySet internally stores this mapping but exposes a `Set`-like interface for easy iteration and JS interop.
 */
@JsExport
@Serializable
data class EntitySet<T : WithId> internal constructor(
    @SerialName("e") private val entities: Set<T>
) {
    @JsName("of")
    constructor(vararg items: T) : this(items.toSet())
    internal constructor(items: List<T>) : this(items.toSet())

    /**
     * Objects are stored here associated with their IDs for fast access
     */
    private val internalIdMap = entities.associateBy { it.id }

    internal val ids = entities.map { it.id }.toSet()
    internal val size = entities.size

    internal fun toTypedArray() = entities.toTypedArray()

    internal operator fun contains(id: Id) = id in internalIdMap
    operator fun get(id: Id) = internalIdMap[id]
    internal operator fun minus(id: Id) = EntitySet(entities.filterNot { it.id == id }.toSet())
    internal operator fun plus(newItem: T) = EntitySet(
        if (newItem.id in internalIdMap) entities.filterNot { it.id == newItem.id }.toSet() + newItem
        else entities + newItem
    )

    internal operator fun plus(newItems: Iterable<T>) = EntitySet(
        entities.filterNot { newItems.any { newNode -> it.id == newNode.id } }.toSet() + newItems
    )

    internal operator fun plus(newItems: Array<T>) = EntitySet(
        entities.filterNot { newItems.any { newNode -> it.id == newNode.id } }.toSet() + newItems
    )

    internal fun isEmpty() = entities.isEmpty()
    internal fun isNotEmpty() = entities.isNotEmpty()

    internal inline fun all(predicate: (T) -> Boolean) = entities.all(predicate)
    internal inline fun any(predicate: (T) -> Boolean) = entities.any(predicate)
    internal inline fun none(predicate: (T) -> Boolean) = entities.none(predicate)

    internal inline fun filter(predicate: (T) -> Boolean) = EntitySet(entities.filter(predicate))
    internal inline fun find(predicate: (T) -> Boolean) = entities.find(predicate)
    internal inline fun <R> flatMap(transform: (T) -> Iterable<R>) = entities.flatMap(transform).toTypedArray()
    fun <R> map(transform: (T) -> R) = entities.map(transform).toTypedArray()
    internal inline fun <U : WithId> mapToNewEntitySet(transform: (T) -> U) = EntitySet(entities.map(transform))
    internal inline fun <R : Comparable<R>> minOf(selector: (T) -> R) = entities.minOf(selector)
}
