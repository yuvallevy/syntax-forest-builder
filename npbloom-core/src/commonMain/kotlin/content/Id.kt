package content

typealias Id = String
typealias IdMap<T> = Map<Id, T>
