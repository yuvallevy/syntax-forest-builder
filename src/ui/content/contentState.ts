import {
  Id, StringSlice, Sentence, NodeIndicatorInPlot, PlotIndex
} from '../../content/types';
import {
  adoptNodesInTree, disownNodesInTree,
  InsertedNode, insertNodeIntoTree, transformNodeInTree, transformNodesInTree
} from '../../content/unpositioned/manipulation';
import { deleteNodesInPlot, transformNodesInPlot } from '../../content/unpositioned/plotManipulation';
import UndoRedoHistory, { ApplyActionFunc, applyToHistory, redo, ReverseActionFunc, undo, UndoableActionCommon } from '../../util/UndoRedoHistory';
import { handleLocalSentenceChange } from './editNodes';
import { changeAt, insertAt, omitKey, removeAt } from '../../util/objTransforms';
import {
  isTerminal, PlotCoordsOffset, UnpositionedPlot, UnpositionedTree
} from '../../content/unpositioned/types';

/**
 * Represents an action taken by the user.
 * To integrate with the undo/redo system, each action is translated into an undoable "state change",
 * which is subsequently applied in a reversible fashion.
 */
export type ContentAction =
  | { type: 'addPlot' }
  | { type: 'deletePlot', plotIndex: PlotIndex }
  | { type: 'insertNode', plotIndex: PlotIndex, treeId: Id, newNodeId: Id, newNode: InsertedNode }
  | { type: 'deleteNodes', plotIndex: PlotIndex, nodeIndicators: NodeIndicatorInPlot[] }
  | { type: 'adoptNodes', plotIndex: PlotIndex, treeId: Id, adoptingNodeId: Id, adoptedNodeIds: Id[] }
  | { type: 'disownNodes', plotIndex: PlotIndex, treeId: Id, disowningNodeId: Id, disownedNodeIds: Id[] }
  | { type: 'moveNodes', plotIndex: PlotIndex, nodeIndicators: NodeIndicatorInPlot[], dx: number, dy: number }
  | { type: 'resetNodePositions', plotIndex: PlotIndex, nodeIndicators: NodeIndicatorInPlot[] }
  | { type: 'setNodeLabel', plotIndex: PlotIndex, nodeIndicator: NodeIndicatorInPlot, newLabel: string }
  | { type: 'setTriangle', plotIndex: PlotIndex, nodeIndicators: NodeIndicatorInPlot[], triangle: boolean }
  | { type: 'setSentence', plotIndex: PlotIndex, treeId: Id, newSentence: Sentence, oldSelectedSlice: StringSlice }
  | { type: 'addTree', plotIndex: PlotIndex, newTreeId: Id, offset: PlotCoordsOffset }
  | { type: 'removeTree', plotIndex: PlotIndex, treeId: Id }
;

/**
 * Represents a change in state as a result of an action done by the user.
 * Each state change includes information about the state before it so it can be easily undone,
 * and each action by the user is translated into a state change so that undo/redo can work smoothly.
 */
type ContentChange = (
  | { type: 'setPlot', plotIndex: PlotIndex, old: UnpositionedPlot, new: UnpositionedPlot }
  | { type: 'addPlot', newPlotIndex: PlotIndex, newPlot: UnpositionedPlot }
  | { type: 'removePlot', plotIndex: PlotIndex, removedPlot: UnpositionedPlot }
  | { type: 'setTree', plotIndex: PlotIndex, treeId: Id, old: UnpositionedTree, new: UnpositionedTree }
  | { type: 'addTree', plotIndex: PlotIndex, newTreeId: Id, newTree: UnpositionedTree }
  | { type: 'removeTree', plotIndex: PlotIndex, treeId: Id, removedTree: UnpositionedTree }
);

type UndoableContentChange = UndoableActionCommon & ContentChange;

/**
 * Translates a user action into a state change, which can later be undone.
 */
const makeUndoable = (state: ContentState) => (action: ContentAction): ContentChange => {
  switch (action.type) {
    case 'addPlot': {
      return {
        type: 'addPlot',
        newPlotIndex: state.plots.length,
        newPlot: { trees: {} },
      };
    }
    case 'deletePlot': {
      return {
        type: 'removePlot',
        plotIndex: action.plotIndex,
        removedPlot: state.plots[action.plotIndex],
      }
    }
    case 'insertNode': {
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId: action.treeId,
        old: state.plots[action.plotIndex].trees[action.treeId],
        new: insertNodeIntoTree(action.newNode)(action.newNodeId)(state.plots[action.plotIndex].trees[action.treeId]),
      };
    }
    case 'deleteNodes': {
      return {
        type: 'setPlot',
        plotIndex: action.plotIndex,
        old: state.plots[action.plotIndex],
        new: deleteNodesInPlot(action.nodeIndicators)(state.plots[action.plotIndex]),
      };
    }
    case 'adoptNodes': {
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId: action.treeId,
        old: state.plots[action.plotIndex].trees[action.treeId],
        new: adoptNodesInTree(action.adoptingNodeId, action.adoptedNodeIds)(
          state.plots[action.plotIndex].trees[action.treeId]),
      };
    }
    case 'disownNodes': {
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId: action.treeId,
        old: state.plots[action.plotIndex].trees[action.treeId],
        new: disownNodesInTree(action.disowningNodeId, action.disownedNodeIds)(
          state.plots[action.plotIndex].trees[action.treeId]),
      };
    }
    case 'moveNodes': {
      return {
        type: 'setPlot',
        plotIndex: action.plotIndex,
        old: state.plots[action.plotIndex],
        new: transformNodesInPlot(node => ({
          ...node,
          offset: { dTreeX: node.offset.dTreeX + action.dx, dTreeY: node.offset.dTreeY + action.dy }
        }))(action.nodeIndicators)(state.plots[action.plotIndex]),
      }
    }
    case 'resetNodePositions': {
      return {
        type: 'setPlot',
        plotIndex: action.plotIndex,
        old: state.plots[action.plotIndex],
        new: transformNodesInPlot(node => ({ ...node, offset: { dTreeX: 0, dTreeY: 0 } }))(action.nodeIndicators)(
          state.plots[action.plotIndex]),
      };
    }
    case 'setNodeLabel': {
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId: action.nodeIndicator.treeId,
        old: state.plots[action.plotIndex].trees[action.nodeIndicator.treeId],
        new: transformNodeInTree(node => ({ ...node, label: action.newLabel }))(action.nodeIndicator.nodeId)(
          state.plots[action.plotIndex].trees[action.nodeIndicator.treeId]),
      };
    }
    case 'setTriangle': {
      const treeId = action.nodeIndicators[0].treeId;  // TODO: Use all trees
      const nodeIds = action.nodeIndicators
        .filter(nodeIndicator => nodeIndicator.treeId === treeId).map(nodeIndicator => nodeIndicator.nodeId);
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId,
        old: state.plots[action.plotIndex].trees[treeId],
        new: transformNodesInTree(node => isTerminal(node) ? { ...node, triangle: action.triangle } : node)(nodeIds)(
          state.plots[action.plotIndex].trees[treeId]),
      };
    }
    case 'setSentence': {
      return {
        type: 'setTree',
        plotIndex: action.plotIndex,
        treeId: action.treeId,
        old: state.plots[action.plotIndex].trees[action.treeId],
        new: handleLocalSentenceChange(action.newSentence, action.oldSelectedSlice)(
          state.plots[action.plotIndex].trees[action.treeId]),
      };
    }
    case 'addTree': {
      return {
        type: 'addTree',
        plotIndex: action.plotIndex,
        newTreeId: action.newTreeId,
        newTree: {
          nodes: {},
          sentence: '',
          offset: action.offset,
        },
      };
    }
    case 'removeTree': {
      return {
        ...action,
        removedTree: state.plots[action.plotIndex].trees[action.treeId],
      };
    }
  }
};

const applyUndoableAction: ApplyActionFunc<UndoableContentChange, ContentState> = action => state => {
  switch (action.type) {
    case 'setPlot': {
      return {
        ...state,
        plots: changeAt(state.plots, action.plotIndex, action.new),
      };
    }
    case 'addPlot': {
      return {
        ...state,
        plots: insertAt(state.plots, action.newPlotIndex, action.newPlot),
      };
    }
    case 'removePlot': {
      return {
        ...state,
        plots: removeAt(state.plots, action.plotIndex),
      }
    }
    case 'setTree': {
      return {
        ...state,
        plots: changeAt(state.plots, action.plotIndex, {
          ...state.plots[action.plotIndex],
          trees: {
            ...state.plots[action.plotIndex].trees,
            [action.treeId]: action.new,
          },
        }),
      };
    }
    case 'addTree': {
      return {
        ...state,
        plots: changeAt(state.plots, action.plotIndex, {
          ...state.plots[action.plotIndex],
          trees: {
            ...state.plots[action.plotIndex].trees,
            [action.newTreeId]: action.newTree,
          },
        }),
      };
    }
    case 'removeTree': {
      return {
        ...state,
        plots: changeAt(state.plots, action.plotIndex, {
          ...state.plots[action.plotIndex],
          trees: omitKey(state.plots[action.plotIndex].trees, action.treeId)
        }),
      };
    }
    default:
      return state;
  }
};

const reverseUndoableAction: ReverseActionFunc<UndoableContentChange> = action => {
  switch (action.type) {
    case 'setPlot': {
      return {
        ...action,
        old: action.new,
        new: action.old,
      };
    }
    case 'addPlot': {
      return {
        ...action,
        type: 'removePlot',
        plotIndex: action.newPlotIndex,
        removedPlot: action.newPlot || { trees: {} },
      };
    }
    case 'removePlot': {
      return {
        ...action,
        type: 'addPlot',
        newPlotIndex: action.plotIndex,
        newPlot: action.removedPlot,
      }
    }
    case 'setTree': {
      return {
        ...action,
        old: action.new,
        new: action.old,
      };
    }
    case 'addTree': {
      return {
        ...action,
        type: 'removeTree',
        plotIndex: action.plotIndex,
        treeId: action.newTreeId,
        removedTree: action.newTree,
      };
    }
    case 'removeTree': {
      return {
        ...action,
        type: 'addTree',
        plotIndex: action.plotIndex,
        newTreeId: action.treeId,
        newTree: action.removedTree,
      };
    }
    default:
      return action;
  }
};

export type ContentState = {
  plots: UnpositionedPlot[];
};

export type UndoableContentState = UndoRedoHistory<UndoableContentChange, ContentState>;

const initialState: ContentState = {
  plots: [
    {
      trees: {},
    },
  ],
};

export const initialContentState: UndoableContentState = {
  current: initialState,
  undoStack: [],
  redoStack: [],
};

export const contentReducer = (state: UndoableContentState, action: ContentAction | { type: 'undo' } | { type: 'redo' }): UndoableContentState =>
  action.type === 'undo' ? undo(applyUndoableAction)(reverseUndoableAction)(state)
    : action.type === 'redo' ? redo(applyUndoableAction)(state)
    : applyToHistory(applyUndoableAction)({ ...makeUndoable(state.current)(action), timestamp: new Date() })(state);
