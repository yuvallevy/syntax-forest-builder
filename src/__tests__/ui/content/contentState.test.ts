import { describe, expect, it } from 'vitest';
import { ContentAction, ContentState, contentReducer, UndoableContentState } from '../../../ui/content/contentState';

describe('content state', () => {
  const initialState: ContentState = {
    plots: [
      {
        trees: {
          'aa': {
            nodes: {
              'a': { label: 'N', offset: { dTreeX: 0, dTreeY: 0 }, slice: [0, 4], triangle: false },
              'b': { label: 'N', offset: { dTreeX: 1, dTreeY: 10 }, slice: [5, 10], triangle: false },
            },
            offset: { dPlotX: 0, dPlotY: 0 },
            sentence: 'tree state',
          },
          'zz': {
            nodes: {
              'w': { label: 'NP', offset: { dTreeX: -1, dTreeY: 5 }, children: ['x'] },
              'x': { label: 'N', offset: { dTreeX: 1, dTreeY: 8 }, slice: [0, 5], triangle: false },
              'y': { label: 'V', offset: { dTreeX: 0, dTreeY: 0 }, slice: [6, 10], triangle: false },
            },
            offset: { dPlotX: 60, dPlotY: 0 },
            sentence: 'nodes rock',
          },
        }
      }
    ],
  };

  const undoableInitialState: UndoableContentState = {
    current: initialState,
    undoStack: [],
    redoStack: [],
  };

  const testCases: [string, ContentAction][] = [
    [
      'inserts a node into a tree',
      { type: 'insertNode', plotIndex: 0, treeId: 'aa', newNodeId: 'c', newNode: { label: 'NP', targetChildIds: ['a', 'b'] } }
    ],
    [
      'deletes a node from a tree',
      { type: 'deleteNodes', plotIndex: 0, nodeIndicators: [{ treeId: 'aa', nodeId: 'b' }] }
    ],
    [
      'deletes two nodes from two different trees',
      { type: 'deleteNodes', plotIndex: 0, nodeIndicators: [{ treeId: 'aa', nodeId: 'b' }, { treeId: 'zz', nodeId: 'x' }] }
    ],
    [
      'sets a node as a child of another',
      { type: 'adoptNodes', plotIndex: 0, treeId: 'zz', adoptingNodeId: 'x', adoptedNodeIds: ['y'] }
    ],
    [
      'removes a connection between two nodes',
      { type: 'disownNodes', plotIndex: 0, treeId: 'zz', disowningNodeId: 'w', disownedNodeIds: ['x'] }
    ],
    [
      'moves a node in a tree',
      { type: 'moveNodes', plotIndex: 0, nodeIndicators: [{ treeId: 'aa', nodeId: 'a' }], dx: 1, dy: -4 }
    ],
    [
      'resets the positions of two nodes in two different trees',
      { type: 'resetNodePositions', plotIndex: 0, nodeIndicators: [{ treeId: 'aa', nodeId: 'b' }, { treeId: 'zz', nodeId: 'w' }] }
    ],
    [
      'sets the label of a node',
      { type: 'setNodeLabel', plotIndex: 0, nodeIndicator: { treeId: 'aa', nodeId: 'a' }, newLabel: 'NP' }
    ],
    [
      'sets whether a terminal node connects to its slice with a triangle',
      { type: 'setTriangle', plotIndex: 0, nodeIndicators: [{ treeId: 'aa', nodeId: 'b' }], triangle: true }
    ],
    [
      'reacts to a change in the sentence',
      { type: 'setSentence', plotIndex: 0, treeId: 'aa', newSentence: 'tee state', oldSelectedSlice: [2, 2] }
    ],
    [
      'adds a tree to a plot',
      { type: 'addTree', plotIndex: 0, newTreeId: 'zz', offset: { dPlotX: 105, dPlotY: 88 } }
    ],
    [
      'removes a tree from a plot',
      { type: 'removeTree', plotIndex: 0, treeId: 'aa' }
    ],
  ];

  it.each(testCases)('%s', (_, action) => expect(contentReducer(undoableInitialState, action).current).toMatchSnapshot());
});
