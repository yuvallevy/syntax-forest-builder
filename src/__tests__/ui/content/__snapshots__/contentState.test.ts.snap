// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`content state > adds a tree to a plot 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                0,
                4,
              ],
              "triangle": false,
            },
            "b": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 10,
              },
              "slice": [
                5,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tree state",
        },
        "zz": {
          "nodes": {},
          "offset": {
            "dPlotX": 105,
            "dPlotY": 88,
          },
          "sentence": "",
        },
      },
    },
  },
}
`;

exports[`content state > deletes a node from a tree 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                0,
                4,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tree state",
        },
        "zz": {
          "nodes": {
            "w": {
              "children": [
                "x",
              ],
              "label": "NP",
              "offset": {
                "dTreeX": -1,
                "dTreeY": 5,
              },
            },
            "x": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 8,
              },
              "slice": [
                0,
                5,
              ],
              "triangle": false,
            },
            "y": {
              "label": "V",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                6,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 60,
            "dPlotY": 0,
          },
          "sentence": "nodes rock",
        },
      },
    },
  },
}
`;

exports[`content state > deletes two nodes from two different trees 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                0,
                4,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tree state",
        },
        "zz": {
          "nodes": {
            "w": {
              "formerDescendants": {
                "x": {
                  "label": "N",
                  "offset": {
                    "dTreeX": 1,
                    "dTreeY": 8,
                  },
                  "slice": [
                    0,
                    5,
                  ],
                  "triangle": false,
                },
              },
              "label": "NP",
              "offset": {
                "dTreeX": -1,
                "dTreeY": 5,
              },
            },
            "y": {
              "label": "V",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                6,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 60,
            "dPlotY": 0,
          },
          "sentence": "nodes rock",
        },
      },
    },
  },
}
`;

exports[`content state > inserts a node into a tree 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                0,
                4,
              ],
              "triangle": false,
            },
            "b": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 10,
              },
              "slice": [
                5,
                10,
              ],
              "triangle": false,
            },
            "c": {
              "children": [
                "a",
                "b",
              ],
              "label": "NP",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tree state",
        },
        "zz": {
          "nodes": {
            "w": {
              "children": [
                "x",
              ],
              "label": "NP",
              "offset": {
                "dTreeX": -1,
                "dTreeY": 5,
              },
            },
            "x": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 8,
              },
              "slice": [
                0,
                5,
              ],
              "triangle": false,
            },
            "y": {
              "label": "V",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                6,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 60,
            "dPlotY": 0,
          },
          "sentence": "nodes rock",
        },
      },
    },
  },
}
`;

exports[`content state > moves a node in a tree 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": -4,
              },
              "slice": [
                0,
                4,
              ],
              "triangle": false,
            },
            "b": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 10,
              },
              "slice": [
                5,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tree state",
        },
        "zz": {
          "nodes": {
            "w": {
              "children": [
                "x",
              ],
              "label": "NP",
              "offset": {
                "dTreeX": -1,
                "dTreeY": 5,
              },
            },
            "x": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 8,
              },
              "slice": [
                0,
                5,
              ],
              "triangle": false,
            },
            "y": {
              "label": "V",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                6,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 60,
            "dPlotY": 0,
          },
          "sentence": "nodes rock",
        },
      },
    },
  },
}
`;

exports[`content state > reacts to a change in the sentence 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                0,
                3,
              ],
              "triangle": false,
            },
            "b": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 10,
              },
              "slice": [
                4,
                9,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tee state",
        },
        "zz": {
          "nodes": {
            "w": {
              "children": [
                "x",
              ],
              "label": "NP",
              "offset": {
                "dTreeX": -1,
                "dTreeY": 5,
              },
            },
            "x": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 8,
              },
              "slice": [
                0,
                5,
              ],
              "triangle": false,
            },
            "y": {
              "label": "V",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                6,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 60,
            "dPlotY": 0,
          },
          "sentence": "nodes rock",
        },
      },
    },
  },
}
`;

exports[`content state > removes a connection between two nodes 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                0,
                4,
              ],
              "triangle": false,
            },
            "b": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 10,
              },
              "slice": [
                5,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tree state",
        },
        "zz": {
          "nodes": {
            "w": {
              "formerDescendants": {
                "x": {
                  "label": "N",
                  "offset": {
                    "dTreeX": 1,
                    "dTreeY": 8,
                  },
                  "slice": [
                    0,
                    5,
                  ],
                  "triangle": false,
                },
              },
              "label": "NP",
              "offset": {
                "dTreeX": -1,
                "dTreeY": 5,
              },
            },
            "x": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 8,
              },
              "slice": [
                0,
                5,
              ],
              "triangle": false,
            },
            "y": {
              "label": "V",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                6,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 60,
            "dPlotY": 0,
          },
          "sentence": "nodes rock",
        },
      },
    },
  },
}
`;

exports[`content state > removes a tree from a plot 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "zz": {
          "nodes": {
            "w": {
              "children": [
                "x",
              ],
              "label": "NP",
              "offset": {
                "dTreeX": -1,
                "dTreeY": 5,
              },
            },
            "x": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 8,
              },
              "slice": [
                0,
                5,
              ],
              "triangle": false,
            },
            "y": {
              "label": "V",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                6,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 60,
            "dPlotY": 0,
          },
          "sentence": "nodes rock",
        },
      },
    },
  },
}
`;

exports[`content state > resets the positions of two nodes in two different trees 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                0,
                4,
              ],
              "triangle": false,
            },
            "b": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                5,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tree state",
        },
        "zz": {
          "nodes": {
            "w": {
              "children": [
                "x",
              ],
              "label": "NP",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
            },
            "x": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 8,
              },
              "slice": [
                0,
                5,
              ],
              "triangle": false,
            },
            "y": {
              "label": "V",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                6,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 60,
            "dPlotY": 0,
          },
          "sentence": "nodes rock",
        },
      },
    },
  },
}
`;

exports[`content state > sets a node as a child of another 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                0,
                4,
              ],
              "triangle": false,
            },
            "b": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 10,
              },
              "slice": [
                5,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tree state",
        },
        "zz": {
          "nodes": {
            "w": {
              "children": [
                "x",
              ],
              "label": "NP",
              "offset": {
                "dTreeX": -1,
                "dTreeY": 5,
              },
            },
            "x": {
              "children": [
                "y",
              ],
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
            },
            "y": {
              "label": "V",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                6,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 60,
            "dPlotY": 0,
          },
          "sentence": "nodes rock",
        },
      },
    },
  },
}
`;

exports[`content state > sets the label of a node 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "NP",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                0,
                4,
              ],
              "triangle": false,
            },
            "b": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 10,
              },
              "slice": [
                5,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tree state",
        },
        "zz": {
          "nodes": {
            "w": {
              "children": [
                "x",
              ],
              "label": "NP",
              "offset": {
                "dTreeX": -1,
                "dTreeY": 5,
              },
            },
            "x": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 8,
              },
              "slice": [
                0,
                5,
              ],
              "triangle": false,
            },
            "y": {
              "label": "V",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                6,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 60,
            "dPlotY": 0,
          },
          "sentence": "nodes rock",
        },
      },
    },
  },
}
`;

exports[`content state > sets whether a terminal node connects to its slice with a triangle 1`] = `
{
  "plots": {
    "plot": {
      "trees": {
        "aa": {
          "nodes": {
            "a": {
              "label": "N",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                0,
                4,
              ],
              "triangle": false,
            },
            "b": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 10,
              },
              "slice": [
                5,
                10,
              ],
              "triangle": true,
            },
          },
          "offset": {
            "dPlotX": 0,
            "dPlotY": 0,
          },
          "sentence": "tree state",
        },
        "zz": {
          "nodes": {
            "w": {
              "children": [
                "x",
              ],
              "label": "NP",
              "offset": {
                "dTreeX": -1,
                "dTreeY": 5,
              },
            },
            "x": {
              "label": "N",
              "offset": {
                "dTreeX": 1,
                "dTreeY": 8,
              },
              "slice": [
                0,
                5,
              ],
              "triangle": false,
            },
            "y": {
              "label": "V",
              "offset": {
                "dTreeX": 0,
                "dTreeY": 0,
              },
              "slice": [
                6,
                10,
              ],
              "triangle": false,
            },
          },
          "offset": {
            "dPlotX": 60,
            "dPlotY": 0,
          },
          "sentence": "nodes rock",
        },
      },
    },
  },
}
`;
